File: Kakera Project/code-1.py
#!/usr/bin/env python3

#==============================================#
#         Created By: Svess#8004               #
#  Last Modification:  2021-03-03 10:32 UTC+0  #
#==============================================#

# Imports
from collections import defaultdict as ddict

# Initialising a default dict to store the reactions
dct = ddict(int)

# List of files that the program will read through
# Should be .txt files
# We recommend compiling your data using DiscordChatExporter
files = [

# Dataset 1 
# 2021-03-15 22:00 UTC - 2021-03-31 07:26 UTC
# After Light kakera was added

"resources-1/flood-1.txt",
"resources-1/flood-2.txt",
"resources-1/flood-3.txt",
"resources-1/flood-4.txt",
"resources-1/flood-5.txt",
"resources-1/flood-6.txt",
"resources-1/flood-7.txt",
"resources-1/flood-8.txt",
"resources-1/flood-9.txt",
"resources-1/flood-husbando.txt",
"resources-1/flood-waifu.txt",
"resources-1/marry-roulette.txt"
]

sm = 0 # Counter for the number of Light kakera

# IDs of the Mudae Bots that counld be used
# You can add more if you have a different 
# Maid/Butler in the server your data is from
MudaDict = {
"Mudae":"Mudae#0807",
"Mudamaid2":"Mudamaid2#2147"
}

# Runs through all the desired files
for File in files:
  searchfile = open(File, "r") # Opening the next File, read only
  lastLine = '' # Initialising a last line to be able to find the reactions later

  # Reads through each line
  for line in searchfile:
    # Checkin if a Message by your specific MudaBOT is a response to a kakeraL reaction
    if ":kakeraL:breaks down into:" in line and MudaDict["Mudamaid2"] in lastLine:
      sm+=1 # Adds to the Light kakera counter
      # Strips the line into the kakera emotes and then
      # splits the line into a list of the kakera broken
      # down from the kakeraL reaction
      lst = list(line[26:-1].split(': =>')[0].split(':+:'))
      # Adding the kakera breakdown to the dict
      for i in lst:
        dct[i]+=1

    lastLine = line # Updates the lastLine variable

  searchfile.close() # Closing the read File so it's ready for the next one

# Summing the number of kakera that the
# Light kakera broke down into
dctsm = 0

for i in dct:
  dctsm += dct[i]

print(dct) # Prints out the raw data

print('\nPurple:', dct['kakeraP'], '\nBlue:', dct['kakera '], '\nTeal:', dct['kakeraT'], 
'\nGreen:', dct['kakeraG'], '\nYellow:', dct['kakeraY'], '\nOrange:', dct['kakeraO'], 
'\nRed:', dct['kakeraR'], '\nRainbow:', dct['kakeraW'], '\nFor a total of',
 dctsm, 'kakera\nBroken down from', sm, 'Light kakera')

File: Kakera Project/code-1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class InfoKakera {

	public static void main(String[] args) {
		
//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**
		String[] fileNames = { //all file locations and names here in this format
"src/resources-1/flood-1.txt",
"src/resources-1/flood-2.txt",
"src/resources-1/flood-3.txt",
"src/resources-1/flood-4.txt",
"src/resources-1/flood-5.txt",
"src/resources-1/flood-6.txt",
"src/resources-1/flood-7.txt",
"src/resources-1/flood-8.txt",
"src/resources-1/flood-9.txt",
"src/resources-1/flood-husbando.txt",
"src/resources-1/flood-waifu.txt",
"src/resources-1/marry-roulette.txt"
};
		String botName = "Mudamaid2#2147"; //the bot's name in this format
//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**
		
		int[] totalOccurrences = {0, 0, 0, 0, 0, 0, 0, 0, 0};
							  //  P  :  T  G  Y  O  R  W  L
		
		for(String s : fileNames) {
			int[] occurrences = {0, 0, 0, 0, 0, 0, 0, 0, 0};
			//  P  :  T  G  Y  O  R  W  L
			String lastLine = "";
			try(BufferedReader reader = new BufferedReader(new FileReader(s))) {
				String line = reader.readLine();
				while(line != null) {
					int[] lineResults = {0, 0, 0, 0, 0, 0, 0, 0, 0};
									 //  P  :  T  G  Y  O  R  W  L
					if(line.contains(":kakeraL:breaks down into:") && lastLine.contains(botName)) { //if the line contains the desired message
						lineResults = getOccurrences(line.substring(25, line.indexOf('=') - 1)); //trims out the irrelevant info
					}
					
					if(lineResults[8] == 1) {
						for(int i=0;i<lineResults.length;i++) {
							occurrences[i] += lineResults[i];
						}
					}
					lastLine = line; //remember the last line
					line = reader.readLine(); //read and go to the next line
				}
				
				//print results
				/*
				 * System.out.println("Purple: "+occurrences[0]);
				 * System.out.println("Normal: "+occurrences[1]);
				 * System.out.println("Teal: "+occurrences[2]);
				 * System.out.println("Green: "+occurrences[3]);
				 * System.out.println("Yellow: "+occurrences[4]);
				 * System.out.println("Orange: "+occurrences[5]);
				 * System.out.println("Red: "+occurrences[6]);
				 * System.out.println("Rainbow: "+occurrences[7]);
				 * System.out.println("Results were gathered out of "+occurrences[8]+" kakera");
				 */
				
			} catch(IOException e) {
				e.printStackTrace();
			}
			
			for(int i=0;i<9;i++) { //add results from file to total results
				totalOccurrences[i] += occurrences[i];
			}
		}
		
		//print total results
		System.out.println();
		System.out.println("Total results:");
		System.out.println("Purple: "+totalOccurrences[0]);
		System.out.println("Normal: "+totalOccurrences[1]);
		System.out.println("Teal: "+totalOccurrences[2]);
		System.out.println("Green: "+totalOccurrences[3]);
		System.out.println("Yellow: "+totalOccurrences[4]);
		System.out.println("Orange: "+totalOccurrences[5]);
		System.out.println("Red: "+totalOccurrences[6]);
		System.out.println("Rainbow: "+totalOccurrences[7]);
		System.out.println("Results were gathered out of "+totalOccurrences[8]+" light kakera");
		
	}
	
	public static int[] getOccurrences(String line) {
		int[] occurrences = {0, 0, 0, 0, 0, 0, 0, 0, 1};
						//  P  :  T  G  Y  O  R  W  L
		
		String[] types = line.split("\\+");
		
		for(String s : types) {
			switch(s.charAt(7)) {
			case 'P':
				occurrences[0] += 1;
				break;
			case ':':
				occurrences[1] += 1;
				break;
			case 'T':
				occurrences[2] += 1;
				break;
			case 'G':
				occurrences[3] += 1;
				break;
			case 'Y':
				occurrences[4] += 1;
				break;
			case 'O':
				occurrences[5] += 1;
				break;
			case 'R':
				occurrences[6] += 1;
				break;
			case 'W':
				occurrences[7] += 1;
				break;
			default:
				System.out.println("fix");
			}
		}
		
		return occurrences;
	}

}


